runner.go

```diff
@@ -0,0 +1,109 @@
+package main
+
+import (
+	"fmt"
+	"io"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"strings"
+	"sync"
+)
+
+// A Runner runs commands.
+type Runner interface {
+	Run(cmd *exec.Cmd, path string) error
+}
+
+// DumpRunner is a Runner that dumps command execution output to disk.
+type DumpRunner struct {
+	dir string
+}
+
+var _ Runner = (*DumpRunner)(nil)
+
+// NewDumpRunner creates a DumpRunner.
+func NewDumpRunner(dir string) *DumpRunner {
+	return &DumpRunner{
+		dir: dir,
+	}
+}
+
+// Run runs cmd and saves the stdout to path, relative to r.dir. Stderr, if any,
+// goes to path.stderr. Parent directories are created if necessary. Path cannot
+// be empty.
+func (r *DumpRunner) Run(cmd *exec.Cmd, path string) error {
+	if path == "" {
+		return fmt.Errorf("command %q: missing path to output", strings.Join(cmd.Args, " "))
+	}
+
+	basedir := filepath.Join(r.dir, filepath.Dir(path))
+	if err := os.MkdirAll(basedir, 0770); err != nil {
+		return err
+	}
+
+	stdout, err := os.Create(filepath.Join(r.dir, path))
+	if err != nil {
+		return err
+	}
+	defer stdout.Close()
+
+	var pathStderr = path + ".stderr"
+	stderr := &lazyFileWriter{path: filepath.Join(r.dir, pathStderr)}
+	defer stderr.Close()
+
+	cmd.Stdout = io.MultiWriter(filterWriters(cmd.Stdout, stdout)...)
```

Comment:
Yes, this just writes the tar file. However, we could add an option to the runner to automatically untar when desired?
What analysis will we be doing on the tar contents? I'd defer implementing that a bit more until we know what we will be reading from those files.